packages/blade/src/components/Collapsible/CollapsiblePanel.web.tsx

```diff
@@ -0,0 +1,125 @@
+ * Overall on expanding height will change as: 0px -> Actual height -> auto
+ */
+const HEIGHT_EXPANDED = 'auto';
+const HEIGHT_COLLAPSED = '0px';
+
+// TODO: move common styles when implementing native
+const StyledCollapsiblePanel = styled.div<StyledCollapsiblePanelProps>((props) => {
+  const { theme, defaultIsExpanded, isExpanded } = props;
+  return {
+    transitionDuration: castWebType(makeMotionTime(theme.motion.duration.xmoderate)),
+    transitionTimingFunction: castWebType(theme.motion.easing.standard.effective),
+    transitionProperty: 'height, opacity',
+    opacity: isExpanded ? 1 : 0.8,
+    /**
+     * We need height explicitly here for initial styles because the component might be rendered on server,
+     * in which case for expanded items this should be `auto` because we don't know the actual pixel value.
+     *
+     * This is based on `defaultIsExpanded` rather than `isExpanded` because post initial render inline styles take over.
+     * Otherwise, changing `height` both here and in inline styles sometimes causes animation flickers due to styles mismatch.
+     */
+    height: defaultIsExpanded ? HEIGHT_EXPANDED : HEIGHT_COLLAPSED,
+    overflowY: 'hidden',
+  };
+});
+
+const CollapsiblePanel = ({ children }: CollapsiblePanelProps): ReactElement => {
+  const { isExpanded, defaultIsExpanded } = useCollapsibleContext();
+  const collapsiblePanelRef = useRef<HTMLDivElement>(null);
+  const { theme } = useTheme();
+  const { isVisible, isAnimating } = usePresence(isExpanded, {
+    transitionDuration: theme.motion.duration.xmoderate,
+  });
+
+  /**
+   * This effect imperatively updates height to make css transitions work:
+   * - for expanded items: auto height -> actual height -> 0px
+   * - for collapsed items: 0px -> actual height -> auto height
+   * - uses `requestAnimationFrame` to set the styles just before the next repaint
+   */
+  useDidUpdate(() => {
+    const collapsiblePanelElement = collapsiblePanelRef.current;
+
+    if (!collapsiblePanelElement) {
+      return;
+    }
+
+    const actualHeight = collapsiblePanelElement.scrollHeight;
+    if (!isExpanded) {
+      // collapse
+      requestAnimationFrame(() => {
+        collapsiblePanelElement.style.height = makeSize(actualHeight);
+
+        requestAnimationFrame(() => {
+          collapsiblePanelElement.style.height = makeSize(0);
+          // collapsiblePanelElement.style.display = 'none';
+        });
+      });
+    } else {
+      // expand
+      requestAnimationFrame(() => {
+        // collapsiblePanelElement.style.display = 'block';
+        collapsiblePanelElement.style.height = makeSize(0);
+
+        requestAnimationFrame(() => {
+          collapsiblePanelElement.style.height = makeSize(actualHeight);
+
+          /**
+           * After this we want to wait for the animation to finish
+           * before setting the height back to auto
+           */
```

Comment:
Hmmm stray comment? what does this reference to?

Reply:
Oh, I'll add some more clarity on this comment. I meant that after this point the next `useEffect` below takes over. Will also see if `onTransitionEnd` can simplify the code.

Reply:
Added in https://github.com/razorpay/blade/pull/1262/commits/88c4b2f421ebbd2cd2bfed38bb413c9473a13a77